cpp/src/ts/batched_kalman.cu

```diff
@@ -0,0 +1,820 @@
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 4) {
+    batched_kalman_loop_kernel_v2<4><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 5) {
+    batched_kalman_loop_kernel_v2<5><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 6) {
+    batched_kalman_loop_kernel_v2<6><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 7) {
+    batched_kalman_loop_kernel_v2<7><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else if (r == 8) {
+    batched_kalman_loop_kernel_v2<8><<<numBlocks, numThreadsPerBlock, 0>>>(
+      ys, nobs, T.data(), Z.data(), RRT.data(), P0.data(), alpha.data(),
+      num_batches, vs, Fs, sum_logFs);
+  } else {
+    throw std::runtime_error(
+      "ERROR: Currently unsupported number of parameters (r).");
+  }
+
+  CUDA_CHECK(cudaPeekAtLastError());
+  CUDA_CHECK(cudaDeviceSynchronize());
+}  // namespace ML
+
+__global__ void batched_kalman_loglike_kernel(double* d_vs, double* d_Fs,
+                                              double* d_sumLogFs, int nobs,
+                                              int num_batches, double* sigma2,
+                                              double* loglike) {
+  using BlockReduce = cub::BlockReduce<double, 128>;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+
+  int tid = threadIdx.x;
+  int bid = blockIdx.x;
+  int num_threads = blockDim.x;
+  double bid_sigma2 = 0.0;
+  for (int it = 0; it < nobs; it += num_threads) {
+    // vs and Fs are in time-major order (memory layout: column major)
+    int idx = (it + tid) + bid * nobs;
+    double d_vs2_Fs = 0.0;
+    if (idx < nobs * num_batches) {
+      d_vs2_Fs = d_vs[idx] * d_vs[idx] / d_Fs[idx];
+    }
+    __syncthreads();
+    double partial_sum = BlockReduce(temp_storage).Sum(d_vs2_Fs, nobs - it);
+    bid_sigma2 += partial_sum;
+  }
+  if (tid == 0) {
+    bid_sigma2 /= nobs;
+    sigma2[bid] = bid_sigma2;
+    loglike[bid] = -.5 * (d_sumLogFs[bid] + nobs * log(bid_sigma2)) -
+                   nobs / 2. * (log(2 * M_PI) + 1);
+  }
+}
+
+void batched_kalman_loglike(double* d_vs, double* d_Fs, double* d_sumLogFs,
+                            int nobs, int num_batches, double* sigma2,
+                            double* loglike) {
+  // BlockReduce uses 128 threads, so here also use 128 threads.
+  const int num_threads = 128;
+  batched_kalman_loglike_kernel<<<num_batches, num_threads>>>(
+    d_vs, d_Fs, d_sumLogFs, nobs, num_batches, sigma2, loglike);
+  CUDA_CHECK(cudaDeviceSynchronize());
```

Comment:
device-sync is costly, please use `cudaStreamSynchronize()` instead, by passing a custom stream.